### DISCUSSION TITLE
Failure to connect to Kinesis: Failed to init kvs producer. Error: Unable to create Rotating Credential provider.

### DISCUSSION THREAD

**Discussion by Alex-Wenner-FHR** [awslabs/amazon-kinesis-video-streams-producer-sdk-cpp#976] [Q&A] [answered]
I get this log after testing the container locally and publishing it to an ECS cluster.

Without Credentials:
```python
Failed to init kvs producer. Error: Unable to create Rotating Credential provider. Error status: 0x15000022
```

With Credentials: 
```python 
describeStreamCurlHandler(): DescribeStream API response: {"message":"The security token included in the request is invalid."}
```

## Thoughts
It appears that even though there is an IAM role attached to the task definition, that the contianer _needs_ generated credentials. I don't necessarily understand why that is the case... but that is what it seems.

## What I have tried

I have tried the following:
- In my dockerfile, initalizing empty env vars. When the container starts, the first thing that is ran is a bit of code that fetches these credentials and sets them in place of the empty env vars.
```Dockerfile 
ENV AWS_ACCESS_KEY_ID=
ENV AWS_SECRET_ACCESS_KEY=
ENV AWS_SESSION_TOKEN=
```

```python
try:
    response = boto3.session.Session().get_credentials()
    # Get the temporary credentials
    credentials = response
    os.environ["AWS_ACCESS_KEY_ID"] = credentials.access_key
    os.environ["AWS_SECRET_ACCESS_KEY"] = credentials.secret_key
    os.environ["AWS_SESSION_TOKEN"] = credentials.token
    print("Credentials Set.")
except Exception as E:
    raise ValueError(f"Token Generation Failed: {E}")
```
Any thoughts would be appreciated. I am not sure why tokens would be stale since the code is grabbing credentials from the IAM role attached to the ECS resource itself. 

I would suggest that these environment variables are actually optional and that an IAM role itself would just _work_

Tagging @niyatim23 for visibility - you helped me work through my previous issue!

**Comment by niyatim23** [comment]
Hi @Alex-Wenner-FHR, can you please share the permissions you are giving to the IAM role? The error code `0x15000022` is `STATUS_FILE_CREDENTIAL_PROVIDER_OPEN_FILE_FAILED`

**Reply by Alex-Wenner-FHR** [reply]
Sure!
My policy looks like this:
```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Sid": "VisualEditor0",
            "Effect": "Allow",
            "Action": [
                "kinesisvideo:PutMedia",
                "kinesisvideo:UpdateStream",
                "kinesisvideo:GetDataEndpoint",
                "kinesisvideo:UpdateDataRetention",
                "kinesisvideo:DescribeStream",
                "kinesisvideo:CreateStream"
            ],
            "Resource": "arn:aws:kinesisvideo:<REG>:<ACCT>:stream/*/*"
        },
        {
            "Sid": "VisualEditor1",
            "Effect": "Allow",
            "Action": "kinesisvideo:ListStreams",
            "Resource": "*"
        }
    ]
}
```

**Reply by niyatim23** [reply]
Can you please try making a describe call from the [awscli](https://docs.aws.amazon.com/cli/latest/reference/kinesisvideo/describe-stream.html) and see if your credentials are working?
Example:
`aws kinesisvideo describe-stream --stream-arn <Stream ARN> `

**Reply by Alex-Wenner-FHR** [reply]
Yeah, this was a bit challenging for me to do based off the way we have our AWS roles and federations set up.

but what I did do was:
- assume the role in which this policy is attached.
- take those temporary creds and set them locally in my Dockerfile as AWS env access, secret and session token.
- boot up the cpp container locally

This indeed worked. I was seeing video streams come through kinesis video. I think that this emulates/proves what we were trying to test with the awscli!

**Reply by niyatim23** [reply]
Okay, so you mean to say the stream works until the credentials need to be rotated? What credential provider are you using?

**Reply by Alex-Wenner-FHR** [reply]
No, the credentials of that role seem to work locally but not in the cloud... I am not sure what you mean by credential provider. We are just using IAM roles!

**Reply by Alex-Wenner-FHR** [reply]
The main difference here, is that locally I am setting these keys statically:

```Dockerfile
ENV AWS_ACCESS_KEY_ID=<ACCESS>
ENV AWS_SECRET_ACCESS_KEY=<SECRET>
ENV AWS_SESSION_TOKEN=<SESSION>
```
as where in the cloud, this is not ideal. we are trying to do so dynamically based off the role attached to the resources itself so the dockerfile is like seen above

```Dockerfile
ENV AWS_ACCESS_KEY_ID=
ENV AWS_SECRET_ACCESS_KEY=
ENV AWS_SESSION_TOKEN=
```
and then set by this script, on container boot (prior to the gstreamer command)

```python
try:
    response = boto3.session.Session().get_credentials()
    # Get the temporary credentials
    credentials = response
    os.environ["AWS_ACCESS_KEY_ID"] = credentials.access_key
    os.environ["AWS_SECRET_ACCESS_KEY"] = credentials.secret_key
    os.environ["AWS_SESSION_TOKEN"] = credentials.token
    print("Credentials Set.")
except Exception as E:
    raise ValueError(f"Token Generation Failed: {E}")
```

**Reply by niyatim23** [reply]
By credential provider, I mean any of [these](https://github.com/awslabs/amazon-kinesis-video-streams-producer-sdk-cpp/blob/master/docs/auth.md#kvs-sdk-auth-integration). When do these credentials expire? Does the SDK upload to KVS until the credentials expire?

**Reply by Alex-Wenner-FHR** [reply]
I mean I guess, static credential provider. No, it does not upload anything at all... That is why I am confused. 

Just to unpack this a bit further, _why_ is a credential provider needed if this container is deployed on an AWS resource that has an IAM role attached to it, that grants all necessary permissions?

**Reply by Alex-Wenner-FHR** [reply]
It feels like a credential provider would be needed assuming this container is not directly in contact with an IAM role.

Like locally. If I were to run this container locally it makes sense to me to have credentials generated and share them to the container. But I am very confused as to why you have to have credentials while running in the cloud, with a grantable IAM role attached.

Can you explain this to me? Maybe I am just misunderstanding this process and that is what is happening.

**Reply by Alex-Wenner-FHR** [reply]
My biggest caveat here, I cannot hardcode AWS Keys & Secrets as ENV variables in the dockerfile.

The py script above is me trying to grab them and set them discreetly. Any thoughts would be much appreciated.

**Comment by disa6302** [comment, marked as answer]
@Alex-Wenner-FHR ,

The problem doesnt seem to be with the credentials to start with. Based on the status code, it seems that the SDK is unable to access the file to read the credentials from. Ensure the file has read permissions for the SDK to parse the file for credentials.

**Reply by Alex-Wenner-FHR** [reply]
@disa6302 thanks for the follow up! There is no credentials file. I am making use of the AWS env vars which is not working... 

But my big question, is if this container is deployed to an AWS resource that has an IAM role attached, with all the access it needs why is it needed to have Access Key, Secret Key, and Session Token? Wouldn't the IAM role be able to facilitate the federation for that resource?

**Reply by disa6302** [reply]
The SDK supports 3 modes of credential providers at the moment:
1. Static: This requires AWS_ACCESS_KEY and AWS_SECRET_ACCESS_KEY to be set via environment variables or as a property in kvssink
2. File credential provider: This also is static credentials based, however, the purpose of this method is if you have an out of band process that is vending out credentials and these credentials are read from the file by the SDK.
3. IoT credential provider: This uses X.509 certificates and IAM role. More details here: https://docs.aws.amazon.com/kinesisvideostreams/latest/dg/how-iot.html

If you are running it within ECS, I am guessing the credentials are being vended using AWS_CONTAINER_CREDENTIALS_FULL_URI. This is not supported in the SDK at the moment. So there needs to be an out of band way to vend these credentials to the SDK. 

Let me know if I have misunderstood your setup and I can clarify better.

**Reply by Alex-Wenner-FHR** [reply]
@disa6302 okay got it. This is helpful... after some internal conversations, I think we can make option 1 work through static credentials.

Is it on your team's roadmap to make this container/sdk work on other platforms such as ECS, EC2, or Lambda through the use of assuming IAM roles?

It seems like this would be the most ideal approach to granting said container access.

**Reply by disa6302** [reply]
We are looking to incorporate it, but no timelines yet. I will update this discussion once we have some update.

**Reply by Alex-Wenner-FHR** [reply]
Okay, thank you! Fantastic! I apologize... I kind of came into this issue/discussion assuming that IAM roles just _should _work__. I will mark this as answered. Thanks for the help @disa6302 & @niyatim23
